thetis/forcing.py

```diff
@@ -644,6 +682,163 @@ def set_fields(self, time):
+
+        Raises an AssertionError if standard_name is not found.
+        :arg ncfile: netCDF Dataset object
+        :arg standard_name: standard_name to look for
+        :returns: name of the netCDF variable
+        """
+        name = None
+        for var in ncfile.variables.values():
+            if hasattr(var, 'standard_name') and var.standard_name == standard_name:
+                name = var.name
+                break
+        msg = f'Variable {standard_name} not found in {ncfile.filepath()}'
+        assert name is not None, msg
+        return name
+
+    def _get_subset_nodes(self, grid_x, grid_y, target_x, target_y):
+        """
+        Retuns grid nodes that are necessary for intepolating onto target_x,y
+        """
+        orig_shape = grid_x.shape
+        grid_xy = numpy.array((grid_x.ravel(), grid_y.ravel())).T
+        target_xy = numpy.array((target_x.ravel(), target_y.ravel())).T
+        tri = qhull.Delaunay(grid_xy)
+        simplex = tri.find_simplex(target_xy)
+        vertices = numpy.take(tri.simplices, simplex, axis=0)
+        nodes = numpy.unique(vertices.ravel())
+        nodes_x, nodes_y = numpy.unravel_index(nodes, orig_shape)
+
+        return nodes, nodes_x, nodes_y
+
+    def _create_interpolator(self, ncfile):
+        """
+        Create compact interpolator by finding the minimal necessary support
+        """
+        lat_name = self._get_nc_var_name(ncfile, 'latitude')
+        lon_name = self._get_nc_var_name(ncfile, 'longitude')
+        lat1d = ncfile[lat_name][:]
+        lon1d = ncfile[lon_name][:]
+        lat, lon = numpy.meshgrid(lat1d, lon1d, indexing='ij')
+        # find valid mask
+        self.valid_mask = None
+        # read a variable and take inverse of its mask
+        for name, var in ncfile.variables.items():
+            if len(var.shape) == 3:
+                v = var[0, ...]  # read first time index
+                self.valid_mask = ~v.mask
+                break
+        assert self.valid_mask is not None, 'could not determine mask'
+        assert self.valid_mask.shape == lat.shape, 'mask has wrong shape {self.mask.shape} {lat.shape}'
+
+        self.nodes, self.ind_lat, self.ind_lon = self._get_subset_nodes(lat, lon, self.mesh_lonlat[:, 1], self.mesh_lonlat[:, 0])
+        lat_subset = lat[self.ind_lat, self.ind_lon]
+        lon_subset = lon[self.ind_lat, self.ind_lon]
+
+        self.valid_mask = self.valid_mask[self.ind_lat, self.ind_lon]
+
+        # omit land mask
+        lat_subset = lat_subset[self.valid_mask]
+        lon_subset = lon_subset[self.valid_mask]
+
+        grid_lat = lat_subset.ravel()
+        grid_lon = lon_subset.ravel()
+        if numpy.ma.isMaskedArray(grid_lat):
+            grid_lat = grid_lat.filled(0.0)
+        if numpy.ma.isMaskedArray(grid_lon):
+            grid_lon = grid_lon.filled(0.0)
+        grid_latlon = numpy.vstack((grid_lat, grid_lon)).T
+
+        # building 2D interpolator, this can take a long time (minutes)
```

Comment:
Might be worth adding a PETSc event, given that it is a part of the program with non-negligible cost.

Reply:
Good idea. I've now added a bunch of events in forcing/interpolation routines.